src/compiler/builder.ts

```diff
@@ -0,0 +1,528 @@
+/// <reference path="program.ts" />
+
+namespace ts {
+    export interface EmitOutput {
+        outputFiles: OutputFile[];
+        emitSkipped: boolean;
+    }
+
+    export interface EmitOutputDetailed extends EmitOutput {
+        diagnostics: Diagnostic[];
+        sourceMaps: SourceMapData[];
+        emittedSourceFiles: SourceFile[];
+    }
+
+    export interface OutputFile {
+        name: string;
+        writeByteOrderMark: boolean;
+        text: string;
+    }
+
+    export interface ChangedProgramFiles {
+        /** Minimal set of list of files that require emit */
+        readonly filesToEmit: ReadonlyArray<string>;
+        /** File paths of source files changed/added/removed or affected by changed files */
+        readonly changedFiles: ReadonlyArray<string>;
+    }
+
+    export interface Builder {
+        /**
+         * This is the callback when file infos in the builder are updated
+         */
+        onProgramUpdateGraph(program: Program, hasInvalidatedResolution: HasInvalidatedResolution): void;

```

Comment:
i do not think we need `graph` in the name of this function, users do not need to know how we represnet the files.

Reply:
Nit, i would call `program` as `newProgram`.